pkg/pillar/cmd/zedrouter/server.go

```diff
@@ -227,3 +253,238 @@ func (hdl hostnameHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+					w.WriteHeader(http.StatusOK)
+					fmt.Fprint(w, "openssh-key")
+					return
+				}
+			}
+		}
+	}
+	w.WriteHeader(http.StatusNotFound)
+}
+
+// ServeHTTP for userDataHandler process requests for user-data
+func (hdl userDataHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	remoteIP := net.ParseIP(strings.Split(r.RemoteAddr, ":")[0])
+	anStatus := lookupAppNetworkStatusByAppIP(hdl.ctx, remoteIP)
+	if anStatus != nil {
+		appConfig := lookupAppNetworkConfig(hdl.ctx, anStatus.Key())
+		if appConfig == nil {
+			log.Errorf("No AppNetworkConfig for %s",
+				anStatus.Key())
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		userData, err := getCloudInitUserData(hdl.ctx, appConfig)
+		if err != nil {
+			log.Errorf("cannot get userData for %s: %v",
+				anStatus.Key(), err)
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		ud, err := base64.StdEncoding.DecodeString(userData)
+		if err != nil {
+			log.Errorf("cannot decode userData for %s: %v",
+				anStatus.Key(), err)
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		w.Header().Add("Content-Type", "application/octet-stream")
+		w.WriteHeader(http.StatusOK)
+		w.Write(ud)
+		return
+	}
+	w.WriteHeader(http.StatusNotFound)
+}
+
+// ServeHTTP for openstackHandler metadata service
+func (hdl openstackHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	dirname, filename := path.Split(strings.TrimSuffix(r.URL.Path, "/"))
+	dirname = strings.TrimSuffix(dirname, "/")
+	remoteIP := net.ParseIP(strings.Split(r.RemoteAddr, ":")[0])
+	anStatus := lookupAppNetworkStatusByAppIP(hdl.ctx, remoteIP)
+	var hostname string
+	var id string
+	if anStatus != nil {
+		hostname = anStatus.DisplayName
+		id = anStatus.UUIDandVersion.UUID.String()
+	} else {
+		w.WriteHeader(http.StatusNotFound)
+		return
+	}
+	switch filename {
+	case "openstack":
+		w.WriteHeader(http.StatusOK)
+		fmt.Fprintln(w, "latest")
+	case "meta_data.json":
+		appConfig := lookupAppNetworkConfig(hdl.ctx, anStatus.Key())
+		if appConfig == nil {
+			log.Errorf("No AppNetworkConfig for %s",
+				anStatus.Key())
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		keys, err := getSSHPasswords(hdl.ctx, appConfig)
+		if err != nil {
+			log.Errorf("cannot get ssh passwords for %s",
+				anStatus.Key())
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		var keysMap []map[string]string
+		publicKeys := make(map[string]string)
+		for ind, key := range keys {
+			keysMap = append(keysMap, map[string]string{
+				"data": fmt.Sprintf("%s\n", key),
+				"type": "ssh",
+				"name": fmt.Sprintf("key-%d", ind),
+			})
+			publicKeys[fmt.Sprintf("key-%d", ind)] = fmt.Sprintf("%s\n", key)
+		}
+		resp, _ := json.Marshal(map[string]interface{}{
+			"uuid":         id,
+			"hostname":     hostname,
+			"name":         hostname,
+			"launch_index": 0,
+			"keys":         keysMap,
+			"public_keys":  publicKeys,
+		})
+		w.WriteHeader(http.StatusOK)
```

Comment:
Missing setting of content-type to json?

And shouldn't you set content-type for all 2xx responses even if the output is text?